### ISSUE TITLE
1.x Can't figure out from crash report where the bad code is

### ISSUE THREAD

**Issue by kaciula** [ReactiveX/RxJava#5368]
I have 35K crashes in my crashlytics dashboard all pointing to one single stacktrace but I can't find the offending code. I have rxjava all over the code so it's not easy to track. Can someone help me reconstruct the rx chain from the stacktrace?

What I deduce from the stack trace is that onErrorResumeNext() with Observable.empty() is being called and somewhere down the chain there is a first() operator without a default and that is the reason it crashes. But I can't find such a chain in my code. I know that behind the scenes the operators may be different so maybe the stacktrace is pointing to a different operator chain?

I would appreciate any hint as the crash numbers are sky rocketing but I can't reproduce the crash on my devices.

PS: Is there some tool I can use to have a more relevant rxjava stacktrace in my crashlytics?

```
Caused by java.util.NoSuchElementException: Sequence contains no elements
       at rx.internal.operators.OperatorSingle$ParentSubscriber.onCompleted(OperatorSingle.java:115)
       at rx.internal.operators.OperatorTake$1.onCompleted(OperatorTake.java:56)
       at rx.internal.operators.OnSubscribeFilter$FilterSubscriber.onCompleted(OnSubscribeFilter.java:99)
       at rx.observers.SerializedObserver.onCompleted(SerializedObserver.java:176)
       at rx.observers.SerializedSubscriber.onCompleted(SerializedSubscriber.java:64)
       at rx.internal.operators.OnSubscribeConcatMap$ConcatMapSubscriber.drain(OnSubscribeConcatMap.java:246)
       at rx.internal.operators.OnSubscribeConcatMap$ConcatMapSubscriber.innerCompleted(OnSubscribeConcatMap.java:209)
       at rx.internal.operators.OnSubscribeConcatMap$ConcatMapInnerSubscriber.onCompleted(OnSubscribeConcatMap.java:345)
       at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onCompleted(OnSubscribeDoOnEach.java:70)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
       at rx.internal.operators.OperatorOnErrorResumeNextViaFunction$4.onCompleted(OperatorOnErrorResumeNextViaFunction.java:101)
       at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onCompleted(OnSubscribeDoOnEach.java:70)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
       at rx.internal.operators.OnSubscribeRedo$4$1.onCompleted(OnSubscribeRedo.java:321)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:656)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
       at rx.internal.operators.OperatorMerge$InnerSubscriber.onCompleted(OperatorMerge.java:857)
       at rx.internal.operators.OperatorDelay$1$1.call(OperatorDelay.java:59)
       at rx.internal.schedulers.EventLoopsScheduler$EventLoopWorker$2.call(EventLoopsScheduler.java:189)
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:390)
       at java.util.concurrent.FutureTask.run(FutureTask.java:234)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:153)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:267)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1080)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:573)
       at java.lang.Thread.run(Thread.java:841)
```

**Comment by akarnokd** [comment]
Not likely without the code base or enabling `RxJavaHooks.enableAssemblyTracking()` in your app at the end user. You can somewhat deduce a chain of operations from the trace:

concat(map -> doOnEach) -> filter -> take -> single

**Comment by akarnokd** [comment]
Another way that doesn't destroy performance that much is to replace all uses of `single()` with a transformer that allows setting a tag and adds that to the message when an onError is passing by:

```java
Observable.Transformer<T, T> withTag(String tag) {
    return o -> o.single().onErrorResumeNext(e -> {
        if (e instanceof NoSuchElementException) {
           e = new NoSuchElementException(tag + ": " + e.getMessage(), e);
        }
        return Observable.error(e);
    });
}

//
source.op()
.op()
.compose(withTag("Doing business #1"))
.op()
...
```

**Comment by kaciula** [comment]
I don't use single() at all in my code but I do use last() (in just one place) which behind the scenes does takeLast(1).single() so I'm guessing this could be it. And the crashes start from the version I introduced this chain. Here is the pseudocode:

```
List<Observable<?>> observables = new ArrayList<>();
Observable<?> stream1 = ...;
observables.add(stream1);
Observable<?> stream2 = ...;
observables.add(stream2);
....
Observable<?> stream10 = ...;
observables.add(stream10);

Observable.concat(observables).last().subscribe(...);
```

So I am trying to run 10 API calls in sequence and wait until all of them are completed. Is there a more standard way to achieve this? Initially I was running each API call through a concatMap so I would have a chain of 10 concatMap() but for some reason (can't remember why) I changed it to this. It seems more elegant...
